[update] Show warning for legacy packages and fix with update

mikeland86 opened the pull request:
## Summary

Add warning if devbox.json contains legacy packages. Fix them with update command. Update will turn legacy packages to `<pkg>@latest` and resolve them to current hash. This guarantees version will not change. If they run update again, it will update to newest version.

## How was it tested?

<img width="840" alt="image" src="https://github.com/jetpack-io/devbox/assets/544948/fe612548-356a-4ed5-810a-78a3942913d1">

```bash
# modified devbox.json to have `curl` package.
devbox update
```

Inspected devbox.json and devbox.lock and verified it was now `curl@latest`

savil reviewed (commented):
not quite following the code here :(

mikeland86 commented on internal/impl/devbox.go:
```diff
@@ -98,6 +100,16 @@ func Open(path string, writer io.Writer) (*Devbox, error) {
 	)
 	box.lockfile = lock
+
+	if box.HasDeprecatedPackages() && !legacyPackagesWarningHasBeenShown {
+		legacyPackagesWarningHasBeenShown = true
+		ux.Fwarning(
+			os.Stderr, // Always stderr. box.writer should probably always be err.
+			"Your devbox.json contains packages in legacy format. "+
+				"Please run `devbox update` to update your devbox.json.\n",
+		)
```
I'll test, but not sure what we can do about that. Maybe another reason why envrc should only call devbox once?

savil replied:
Even with one shellenv call, we do invoke devbox again in the generated shellenv (i think)

but maybe we can look into that and remove it. IIRC it may be older stuff like `devbox log` which we no longer use or need.

savil replied:
Actually, maybe we do need it since we care about perf for shell-start

mikeland86 commented on internal/impl/update.go:
```diff
@@ -20,19 +21,37 @@ func (d *Devbox) Update(ctx context.Context, pkgs ...string) error {
+			if err := d.Remove(ctx, pkg.Raw); err != nil {
+				return err
+			}
+			if err := d.lockfile.ResolveToCurrentNixpkgCommitHash(
+				pkg.LegacyToVersioned(),
+			); err != nil {
+				return err
+			}
+			if err := d.Add(ctx, pkg.LegacyToVersioned()); err != nil {
```
`ResolveToCurrentNixpkgCommitHash` will resolve the new "LegacyToVersioned" version of the package (i.e. `curl` -> `curl@latest`) to the same hash that was previously being used (either the default or one specified in config).

When `Add` is called, it will resolve to whatever is in lockfile (`Add` does not update version if package is already pinned in lockfile). This behavior mimics `yarn`. So there is no overwriting.

savil replied:
ah, I see. Thanks 🙏 

So, the purpose of Add here to ensure we add the newly resolved package to the nix profile?

mikeland86 replied:
nix profile, devbox.json and any other things we go when we add (create wrappers, plugins, etc)

savil commented on internal/impl/update.go:
```diff
@@ -5,37 +5,48 @@ import (
-		existing := d.lockfile.Packages[pkg]
-		newEntry, err := searcher.Client().Resolve(pkg)
+		existing := d.lockfile.Packages[pkg.Raw]
+		newEntry, err := searcher.Client().Resolve(pkg.Raw)
```
is this related to your comment in the other PR about using input.Raw instead of input.String()?

mikeland86 commented:
@savil I fixed the direnv issue, PTAL.

* I added `--install` flag to shellenv to merge the commands.
* I also added a way to do devbox.Open that skips warnings for telemetry and autocomplete use cases.

mikeland86 commented:
See https://github.com/jetpack-io/devbox/pull/1107/commits/d19224142a12ae9395bb0dac9a56c41b3227d65a for latest changes
